lua/libcairo_text_helper.h

```diff
@@ -0,0 +1,281 @@
+  FcDefaultSubstitute(pat);
+    
+  FcResult result;
+  FcPattern* font_pat = FcFontMatch(config, pat, &result);
+
+  /* Will be freed with font */
+  FcChar8* file = NULL; 
+  FcPatternGetString(font_pat, FC_FILE, 0, &file);
+  
+  if (!file) {
+    /* FIXME: add error handling */
+    return NULL;
+  }
+  
+  FontData *font_data = malloc(sizeof(struct _FontData));
+  
+  /* Load the font */
+  FT_Init_FreeType(&font_data->ft_library);
+  FT_Library_SetLcdFilter(font_data->ft_library, FT_LCD_FILTER_DEFAULT);
+  FT_New_Face(font_data->ft_library, (char *)file, 0, &font_data->ft_face);
+  FT_Set_Char_Size(font_data->ft_face, font_size*64, font_size*64,0,0);
+  
+  /* Store font data for use later */
+  font_data->cairo_ft_face = cairo_font_face_reference(cairo_ft_font_face_create_for_ft_face(font_data->ft_face, 0));
+  font_data->hb_ft_font = hb_ft_font_create(font_data->ft_face, NULL);
+  font_data->hb_ft_face = hb_ft_face_create(font_data->ft_face, NULL);
+  font_data->font_size = font_size;
+  
+  /* Cleanup font config */
+  FcPatternDestroy(font_pat);
+  FcPatternDestroy(pat);
+  FcConfigDestroy(config);
+  /* FIXME: Crashes */
+  /*FcFini(); */
+  
+  return font_data;
+}
+
+void cairo_text_hp_destroy_font(FontData *font)
+{
+  cairo_font_face_destroy(font->cairo_ft_face);
+  cairo_font_face_destroy (font->cairo_ft_face);
+  hb_font_destroy(font->hb_ft_font);
+  hb_face_destroy(font->hb_ft_face);
+  
+  FT_Done_Face (font->ft_face);
+  FT_Done_FreeType(font->ft_library);
+  
+  free(font);
+}
+
+/* 
+ * Direction calls hb_direction_from_string example values are LTR and RTL
+ *   https://harfbuzz.github.io/harfbuzz-hb-common.html#hb-direction-from-string
+ * Script is an ISO 15924 4 character string, "Zyyy" can be used for "Common" and "Zinh"
+ *  for "Inherited". 
+ *  https://harfbuzz.github.io/harfbuzz-hb-common.html#hb-script-from-string
+ * Language is a BCP 47 language tag. eg "en" or "en-US"
+ */
+void cairo_text_hp_intl_show(cairo_t *cr, int x, int y, cairo_text_alignment_t alignment, const char *text, FontData *font,
+                             const char *direction, const char *script, const char *language)
+{  
+  /* FIXME: Support others */
+  hb_direction_t text_direction = hb_direction_from_string(direction, -1);
+  if (text_direction == HB_DIRECTION_INVALID) {
+    text_direction = HB_DIRECTION_LTR;
+  }
+  hb_script_t text_script = hb_script_from_string(script, -1);
+  if (text_script == HB_SCRIPT_UNKNOWN) {
+    text_script = HB_SCRIPT_COMMON;
+  }
+    
+  hb_language_t text_language = hb_language_from_string (language, -1);
+  
+  /* Draw text */
+  /* Create a buffer for harfbuzz to use */
+  hb_buffer_t *buf = hb_buffer_create();
+
+  //alternatively you can use hb_buffer_set_unicode_funcs(buf, hb_glib_get_unicode_funcs());
+  hb_buffer_set_unicode_funcs(buf, hb_glib_get_unicode_funcs());
+  hb_buffer_set_direction(buf, text_direction); /* or LTR */
+  hb_buffer_set_script(buf, text_script); /* see hb-unicode.h */
```

Comment:
Which `text_script` are you using to test out arabic? It should be "Arab" as defined in [`hb-common.h`](https://github.com/harfbuzz/harfbuzz/blob/fd5eb10d3b4cfee1bb10614ff6719378c949492e/src/hb-common.h#L526C17-L526C17). I'm asking because you're referencing `hb-unicode.h` which might be the reason why HarfBuzz is not shaping the arabic example properly.